pkg/volume/util/goroutinemap/goroutinemap.go

```diff
@@ -0,0 +1,252 @@
+	volumeName api.UniqueVolumeName,
+	podName types.UniquePodName,
+	operationFunc func() error) error {
+	grm.lock.Lock()
+	defer grm.lock.Unlock()
+
+	var previousOp operation
+	opExists := false
+	previousOpIndex := -1
+	for i, existingOp := range grm.operations {
+		if existingOp.volumeName != volumeName {
+			// No match, keep searching
+			continue
+		}
+
+		if existingOp.podName != emptyUniquePodName &&
+			podName != emptyUniquePodName &&
+			existingOp.podName != podName {
+			// No match, keep searching
+			continue
+		}
+
+		// Match
+		opExists = true
+		previousOp = existingOp
+		previousOpIndex = i
+		break
+	}
+
+	if opExists {
+		// Operation already exists
+		if previousOp.operationPending {
+			// Operation is pending
+			operationName := getOperationName(volumeName, podName)
+			return common.NewAlreadyExistsError(operationName)
+		}
+
+		if !previousOp.expBackoff.SafeToRetry() {
+			operationName := getOperationName(volumeName, podName)
+			return common.NewExponentialBackoffError(
+				operationName, previousOp.expBackoff)
+		}
+	}
+
+	if opExists {
+		// Update existing operation to mark as pending.
+		grm.operations[previousOpIndex].operationPending = true
+		grm.operations[previousOpIndex].volumeName = volumeName
+		grm.operations[previousOpIndex].podName = podName
+	} else {
+		// Create a new operation
+		grm.operations = append(grm.operations,
+			operation{
+				operationPending: true,
+				volumeName:       volumeName,
+				podName:          podName,
+				expBackoff:       previousOp.expBackoff,
+			})
+	}
+
+	go func() (err error) {
+		// Handle unhandled panics (very unlikely)
+		defer k8sRuntime.HandleCrash()
+		// Handle completion of and error, if any, from operationFunc()
+		defer grm.operationComplete(volumeName, podName, &err)
+		// Handle panic, if any, from operationFunc()
+		operationName := getOperationName(volumeName, podName)
+		defer common.RecoverFromPanic(operationName, &err)
+		return operationFunc()
+	}()
+
+	return nil
+}
+
+func (grm *goRoutineMap) getOperation(
+	volumeName api.UniqueVolumeName,
+	podName types.UniquePodName) (uint, error) {
+
+	for i, op := range grm.operations {
+		if op.volumeName == volumeName &&
+			op.podName == podName {
+			return uint(i), nil
+		}
+	}
+
+	logOperationName := getOperationName(volumeName, podName)
+	return 0, fmt.Errorf("Operation %q not found.", logOperationName)
+}
+
+func (grm *goRoutineMap) deleteOperation(
+	volumeName api.UniqueVolumeName,
+	podName types.UniquePodName) {
+
+	opIndex := -1
+	for i, op := range grm.operations {
+		if op.volumeName == volumeName &&
+			op.podName == podName {
+			opIndex = i
+			break
+		}
+	}
+
+	// Delete index without preserving order
+	grm.operations[opIndex] = grm.operations[len(grm.operations)-1]
+	grm.operations = grm.operations[:len(grm.operations)-1]
+}
+
+func (grm *goRoutineMap) operationComplete(
+	volumeName api.UniqueVolumeName, podName types.UniquePodName, err *error) {
+	defer grm.cond.Signal()
```

Comment:
@smarterclayton @saad-ali SO defer order according to https://blog.golang.org/defer-panic-and-recover is Last-In is First-Out. Which means that what we are doing here as I under stand it is

1) lock the structure
2) Do all the work in the datastructure required.
3) unlock structure
4) Wakeup anything potentially waiting before this routine takes the lock, or while the lock is taken here. There is a window between the defers where we release the lock and then do the signal but in that case the Wait routine should perform the correct action as the length of the slice should be updated.

Saad is that correct? I recommend commenting the desired order of the defers and the reasoning behind them.